rfc-146-production-deploy-access.md

```diff
@@ -0,0 +1,82 @@
+
+There were [attempts at formalising this process around 2016](https://docs.google.com/document/d/1lo1JAkFIeWrIl-7bzkzps8rYi2hscV6IOpDzc8dhFLs/edit#). Unfortunately we didn't go into detail on why 2ndline shifts were required. The theory is that the logic went something like this:
+
+- In the early days of GOV.UK our deployment pipeline and monitoring was immature
+- This meant that debugging issues with deployments often required ssh access
+- In turn, this meant that only people with ssh access were allowed to do deployments
+- Because it's easy to make mistakes when using highly privileged access through ssh, we required people to do a couple of 2ndline shifts to make sure they knew what they were doing
+
+Note that by this logic, the 2ndline requirements are defending against accidental mistakes, not against malicious access. We have separate defences against malicious employees (security clearance, probation etc.).
+
+## Proposal
+
+We should implement a new level of access - "Production Deploy Access". This level of access should include:
+
+- Permission to deploy apps in Jenkins (ideally without Jenkins admin permission)
+- Permission to merge continuously deployed applications
+- Readonly access to AWS, Fastly, logging systems, etc.
+- admin access to GOV.UK Signon (create and edit normal users)
+
+This level of access should not include:
+
+- ssh permission to production or staging
+- admin access to the Deploy Jenkins in production or staging
+- admin or poweruser access to AWS
+- admin access to other systems (e.g. Fastly etc.)
+- superadmin access to GOV.UK Signon (create and edit all user types and edit applications)
+
+Engineers should be granted this level of access once they've got the required level of security clearance and have been working on GOV.UK for at least one full month. Access should be granted at the discretion of the engineer's tech lead. Before approving access, tech leads should ensure that the engineer:
+
+- is aware of our processes and standards around code review
+- understands the responsibilities that releasing code brings with it
+- knows how to roll back to an older release if there are any issues
+- knows how to get help from someone with more access if they need it
+
+These rules should be the same for contractors and civil servants at all levels. Junior and apprentice developers should be given production deploy access once their tech lead feels it's appropriate.
+
+The rules for gaining Production Admin Access should remain the same - engineers will still be required to complete two 2ndline shifts.
+
+## Consequences
+
+We will update the [Rules for Getting Production Access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) to match the outcome of this RFC.
+
+We will create a new GitHub Team for the middle level of access, so we'll have "GOV.UK", "GOV.UK Deploy", and "GOV.UK Production".
```

Comment:
I like the team names, but wonder if these should be consistent with the terminology used earlier, i.e.

```suggestion
We will create a new GitHub Team for the middle level of access, so we'll have "GOV.UK", "GOV.UK Production Deploy", and "GOV.UK Production Admin".
```

(Or use "GOV.UK Deploy" and "GOV.UK Production" throughout this RFC).

Reply:
I sort of agree, but I'm not sure what I might break by renaming the GitHub teams. At least [the Jenkins Access controls](https://github.com/alphagov/govuk-puppet/blob/e5f47f2d4330ff0713145f93747364921ad3e4b9/hieradata_aws/class/production/jenkins.yaml#L28) would have to be updated to include the new team names.

Can you think of anything else? I can shoot for the clearer names, and if it becomes too difficult I guess I can rename them back.